doc/api/async_hooks.md

```diff
@@ -0,0 +1,677 @@
+the `resource` object passed to `init()` it's possible that `destroy()` is
+never called. Causing a memory leak in the application. Of course if 
+the resource doesn't depend on GC then this isn't an issue.
+
+#### `async_hooks.currentId()`
+
+* Returns {number} the `id` of the current execution context. Useful to track when
+  something fires. 
+
+For example:
+
+```js
+console.log(async_hooks.currentId());  // 1 - bootstrap
+fs.open(path, (err, fd) => {
+  console.log(async_hooks.currentId());  // 2 - open()
+}):
+```
+
+It is important to note that the id returned fom `currentId()` is related to
+execution timing. Not causality (which is covered by `triggerId()`). For
+example:
+
+```js
+const server = net.createServer(function onConnection(conn) {
+  // Returns the id of the server, not of the new connection. Because the
+  // on connection callback runs in the execution scope of the server's
+  // MakeCallback().
+  async_hooks.currentId();
+
+}).listen(port, function onListening() {
+  // Returns the id of a TickObject (i.e. process.nextTick()) because all
+  // callbacks passed to .listen() are wrapped in a nextTick().
+  async_hooks.currentId();
+});
+```
+
+#### `async_hooks.triggerId()`
+
+* Returns {number} the id of the resource responsible for calling the callback
+  that is currently being executed.
+
+For example:
+
+```js
+const server = net.createServer(conn => {
+  // Though the resource that caused (or triggered) this callback to
+  // be called was that of the new connection. Thus the return value
+  // of triggerId() is the id of "conn".
+  async_hooks.triggerId();
+
+}).listen(port, () => {
+  // Even though all callbacks passed to .listen() are wrapped in a nextTick()
+  // the callback itself exists because the call to the server's .listen()
+  // was made. So the return value would be the id of the server.
+  async_hooks.triggerId();
+});
+```
+
+## Embedder API
+
+Library developers that handle their own I/O will need to hook into the
+AsyncWrap API so that all the appropriate callbacks are called. To accommodate
+this a JavaScript API is provided.
+
+### `class AsyncEvent()`
+
+The class `AsyncEvent` was designed to be extended by the embedder's async
+resources. Using this users can easily trigger the lifetime events of their
+own resources.
+
+The `init()` hook will trigger when an `AsyncEvent` is instantiated.
+ 
+It is important that before/after calls are unwound
+in the same order they are called. Otherwise an unrecoverable exception
+will be made.
+
+```js
+// AsyncEvent() is meant to be extended. Instantiating a new AsyncEvent() also
+// triggers init(). If triggerId is omitted then currentId() is used.
+const asyncEvent = new AsyncEvent(type[, triggerId]);
+
+// Call before() hooks.
+asyncEvent.emitBefore();
+
+// Call after() hooks.
+asyncEvent.emitAfter();
+
+// Call destroy() hooks.
+asyncEvent.emitDestroy();
+
+// Return the unique id assigned to the AsyncEvent instance.
+asyncEvent.asyncId();
```

Comment:
not a doc issue, but I don't understand why this isn't simply `id()`, which is what its called elsewhere, AFAICT